src/algebra/algebra/basic.lean

```diff
@@ -41,14 +41,27 @@ set_option extends_priority 200 /- control priority of
 `instance [algebra R A] : has_scalar R A` -/
 
 /--
```

Comment:
I really like the idea of using this doc string to explain like this and I very much like what you've written but I think I'd start by putting the emphasis on the fact that there are two ways of defining unital associative algebras.

The more-bundled way:
```lean
variables [comm_semiring R] [semiring A] [algebra R A]
```
The less-bundled way:
```lean
variables [semiring R] [semiring A] [module R A]
variables [smul_comm_class R A A] [is_scalar_tower R A A]
```
And while pointing out that they give exactly the same theory, I'd probably include:
```lean
example [comm_semiring R] [semiring A] [module R A]
  [smul_comm_class R A A] [is_scalar_tower R A A] : algebra R A :=
algebra.of_module smul_mul_assoc mul_smul_comm
```
And with all that out of the way I'd point out that the less-bundled way has the advantage that the `semiring A` can be swapped out for any of:

- `non_unital_non_assoc_semiring A`
- `non_assoc_semiring A`
- `non_unital_semiring A`

so we can get more general classes of algebras this way.

Reply:
Updated, let me know if you think I've got the emphasis wrong.

Reply:
I think the new version is excellent!